README
SYNOPSIS

      use AnyEvent::MQTT;
      my $mqtt = AnyEvent::MQTT->new;
      my $cv = $mqtt->subscribe(topic => '/topic',
                                callback => sub {
                                     my ($topic, $message) = @_;
                                     print $topic, ' ', $message, "\n"
                                   });
      my $qos = $cv->recv; # subscribed, negotiated QoS == $qos
    
      # publish a simple message
      $cv = $mqtt->publish(message => 'simple message',
                              topic => '/topic');
      $cv->recv; # sent
    
      # publish line-by-line from file handle
      $cv =  $mqtt->publish(handle => \*STDIN,
                            topic => '/topic');
      $cv->recv; # sent
    
      # publish from AnyEvent::Handle
      $cv = $mqtt->publish(handle => AnyEvent::Handle->new(my %handle_args),
                           topic => '/topic');
      $cv->recv; # sent

DESCRIPTION

    AnyEvent module for MQTT client.

    IMPORTANT: This is an early release and the API is still subject to
    change.

DISCLAIMER

    This is not official IBM code. I work for IBM but I'm writing this in
    my spare time (with permission) for fun.

    Constructs a new AnyEvent::MQTT object. The supported parameters are:

    host

      The server host. Defaults to 127.0.0.1.

    port

      The server port. Defaults to 1883.

    timeout

      The timeout for responses from the server.

    keep_alive_timer

      The keep alive timer.

    user_name

      The user name for the MQTT broker.

    password

      The password for the MQTT broker.

    tls

      Set flag to enable TLS encryption, Default is no encryption.

    will_topic

      Set topic for will message. Default is undef which means no will
      message will be configured.

    will_qos

      Set QoS for will message. Default is 'at-most-once'.

    will_retain

      Set retain flag for will message. Default is 0.

    will_message

      Set message for will message. Default is the empty message.

    clean_session

      Set clean session flag for connect message. Default is 1.

    client_id

      Sets the client id for the client overriding the default which is
      NetMQTTpmNNNNN where NNNNN is the current process id.

    message_log_callback

      Defines a callback to call on every message.

    on_error

      Defines a callback to call when some error occurs.

      Two parameters are passed to the callback.

          $on_error->($fatal, $message)

      where $fatal is a boolean flag and $message is the error message. If
      the error is fatal, $fatal is true.

    handle_args

        a reference to a list to pass as arguments to the
        L<AnyEvent::Handle> constructor (defaults to
        an empty list reference).

    This method attempts to destroy any resources in the event of a
    disconnection or fatal error.

    This method is used to publish to a given topic. It returns an AnyEvent
    condvar which is notified when the publish is complete (written to the
    kernel or ack'd depending on the QoS level). The parameter hash must
    included at least a topic value and one of:

    message

        with a string value which is published to the topic,

    handle

       the value of which must either be an L<AnyEvent::Handle> or will be
       passed to an L<AnyEvent::Handle> constructor as the C<fh> argument.
       The L<push_read()> method is called on the L<AnyEvent::Handle> with a
       callback that will publish each chunk read to the topic.

    The parameter hash may also keys for:

    qos

        to set the QoS level for published messages (default
        MQTT_QOS_AT_MOST_ONCE),

    handle_args

        a reference to a list to pass as arguments to the
        L<AnyEvent::Handle> constructor in the final case above (defaults to
        an empty list reference), or

    push_read_args

        a reference to a list to pass as the arguments to the
        L<AnyEvent::Handle#push_read> method (defaults to ['line'] to read,
        and subsequently publish, a line at a time.

    Returns a 16-bit number to use as the next message id in a message
    requiring an acknowledgement.

    This method subscribes to the given topic. The parameter hash may
    contain values for the following keys:

    topic

        for the topic to subscribe to (this is required),

    callback

        for the callback to call with messages (this is required),

    qos

        QoS level to use (default is MQTT_QOS_AT_MOST_ONCE),

    cv

        L<AnyEvent> condvar to use to signal the subscription is complete.
        The received value will be the negotiated QoS level.

    This method returns the value of the cv parameter if it was supplied or
    an AnyEvent condvar created for this purpose.

    This method unsubscribes to the given topic. The parameter hash may
    contain values for the following keys:

    topic

        for the topic to unsubscribe from (this is required),

    callback

        for the callback to call with messages (this is optional and currently
        not supported - all callbacks are unsubscribed),

    cv

        L<AnyEvent> condvar to use to signal the unsubscription is complete.

    This method returns the value of the cv parameter if it was supplied or
    an AnyEvent condvar created for this purpose.

    This method starts the connection to the server. It will be called
    lazily when required publish or subscribe so generally is should not be
    necessary to call it directly.

    This method is used to register an AnyEvent::Handle read type method to
    read MQTT messages.

POD ERRORS

    Hey! The above document had some coding errors, which are explained
    below:

    Around line 39:

      Unknown directive: =method

    Around line 123:

      Unknown directive: =method

    Around line 128:

      Unknown directive: =method

    Around line 174:

      Unknown directive: =method

    Around line 179:

      Unknown directive: =method

    Around line 208:

      Unknown directive: =method

    Around line 233:

      Unknown directive: =method

    Around line 239:

      Unknown directive: =method



Changes
Revision history for Perl extension AnyEvent::MQTT.

 1.212810  2021-10-08 15:57:55+01:00 Europe/London

1.212810  2021-10-08 13:40:52+01:00 Europe/London
    - Merge pr #12 - adding AnyEvent tls option. Thanks to @jr-dimedis.

1.202052  2020-07-23 22:29:34+01:00 Europe/London
    - Merge pr #9
    - Merge pr #7 which fixes #2

1.202051  2020-07-23 18:56:03+01:00 Europe/London
    - Fixes issue/pr #10
    - Fixes issue/pr #11

1.202050  2020-07-23 18:53:17+01:00 Europe/London

1.172121  2017-07-31 22:05:49+01:00 Europe/London
    - Fix travis builds (and changelog)

1.172120  2017-07-31 19:35:44+01:00 Europe/London
    - Add client_id option to command line examples
    - Add handle_args which is useful for TLS arguments. Thanks to
      Matt (@mattaltus)

1.142640  2014-09-21 19:14:14+01:00 Europe/London
    - Documentation fixes. (Thanks to Toshio Ito.)

1.142230  2014-08-11 18:32:39+01:00 Europe/London
    - Don't check locale-specific error message to avoid test failure.

1.142010  2014-07-20 13:35:48+01:00 Europe/London
    - Fix message ids large than 16-bits. (Thanks to wnagele on github.)

1.141992  2014-07-18 14:36:30+01:00 Europe/London
    - Fix bad release tar.gz.

1.141991  2014-07-18 14:17:51+01:00 Europe/London
    - Use on_read to avoid broken pipe on eof.
    - Support username and password in connect message.
    - Add -e option to command line for neat one-liners.

1.112340  2011-08-22 20:22:40 Europe/London
    - Support for retained messages with multiple subscriptions to the
      same topic.
    - Fix multiple subscriptions not unsubscribing properly.

1.112320  2011-08-20 23:12:35 Europe/London
    - Support multiple subscriptions to the same topic with different
      callback functions.
    - Disconnect support.

1.111981  2011-07-17 22:54:00 Europe/London
    - Fix PAUSE indexing problem.

1.111980  2011-07-17 22:31:37 Europe/London
    - Add clean disconnect support.
    - Make Gtk2 modules suggested dependencies since only
      anyevent-mqtt-monitor uses them - thx to Roger Light for
      reminding me about this.

1.110880  2011-03-29 22:23:16 Europe/London
    - New script 'anyevent-mqtt-monitor' (requires Gtk2).
    - Support for simple unsubscribe.
    - New callback, 'message_log_callback', to log every incoming and
      outgoing message.
    - Avoid some cyclic references that lead to memory leaks.
    - Use Sub::Name to aid debugging of anonymous subroutines.

1.110390  2011-02-08 23:01:05 Europe/London
    - Add retain flag support.
    - Add missing debian libanyevent-perl dependency.
    - Improve tests to make them easier to debug.

1.110240  2011-01-24 21:19:02 Europe/London
    - Support subscribe method with QoS levels 1 and 2.
    - Support publish method with QoS levels 1 and 2.

1.110220  2011-01-22 15:29:52 Europe/London
    - Initial release.


t/Helpers.pm
use strict;
use warnings;
package t::Helpers;

=head1 NAME

t::Helpers - Perl extension for Helper functions for tests.

=head1 SYNOPSIS

  use Test::More tests => 2;
  use t::Helpers qw/:all/;
  is(test_error(sub { die 'argh' }),
     'argh',
     'died horribly');

  is(test_warn(sub { warn 'danger will robinson' }),
     'danger will robinson',
     'warned nicely');

=head1 DESCRIPTION

Common functions to make test scripts a bit easier to read.

=cut

use base 'Exporter';
use constant {
  DEBUG => $ENV{ANYEVENT_MQTT_TEST_HELPERS_DEBUG}
};
use AnyEvent;
use AnyEvent::Socket;
use AnyEvent::Handle;
use File::Temp qw/tempfile/;
use Test::More;

our %EXPORT_TAGS = ( 'all' => [ qw(
                                   test_error
                                   test_warn
                                   test_output
) ] );
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
our @EXPORT = qw();

=head2 C<test_error($code_ref)>

This method runs the code with eval and returns the error.  It strips
off some common strings from the end of the message including any "at
<file> line <number>" strings and any "(@INC contains: .*)".

=cut

sub test_error {
  my $sub = shift;
  eval { $sub->() };
  my $error = $@;
  if ($error) {
    $error =~ s/\s+at (\S+|\(eval \d+\)(\[[^]]+\])?) line \d+\.?\s*$//g;
    $error =~ s/\s+at (\S+|\(eval \d+\)(\[[^]]+\])?) line \d+\.?\s*$//g;
    $error =~ s/ \(\@INC contains:.*?\)$//;
  }
  return $error;
}

=head2 C<test_warn($code_ref)>

This method runs the code with eval and returns the warning.  It strips
off any "at <file> line <number>" specific part(s) from the end.

=cut

sub test_warn {
  my $sub = shift;
  my $warn;
  local $SIG{__WARN__} = sub { $warn .= $_[0]; };
  eval { $sub->(); };
  die $@ if ($@);
  if ($warn) {
    $warn =~ s/\s+at (\S+|\(eval \d+\)(\[[^]]+\])?) line \d+\.?\s*$//g;
    $warn =~ s/\s+at (\S+|\(eval \d+\)(\[[^]]+\])?) line \d+\.?\s*$//g;
    $warn =~ s/ \(\@INC contains:.*?\)$//;
  }
  return $warn;
}

sub test_output {
  my ($sub, $fh) = @_;
  my ($tmpfh, $tmpfile) = tempfile();
  open my $oldfh, ">&", $fh     or die "Can't dup \$fh: $!";
  open $fh, ">&", $tmpfh or die "Can't dup \$tmpfh: $!";
  $sub->();
  open $fh, ">&", $oldfh or die "Can't dup \$oldfh: $!";
  $tmpfh->flush;
  open my $rfh, '<', $tmpfile;
  local $/;
  undef $/;
  my $c = <$rfh>;
  close $rfh;
  unlink $tmpfile;
  $tmpfh->close;
  return $c;
}

1;


t/MockServer.pm
use strict;
use warnings;
package t::MockServer;

=head1 NAME

t::MockServer - Perl extension for Mock Server using AnyEvent

=head1 SYNOPSIS

  use t::MockServer;
  my $server = t::MockServer->new([ [ recv => 'hello', 'received hello' ],
                                    [ send => 'test', 'sent test' ] ], [...]);

=head1 DESCRIPTION

Common functions to make test scripts a bit easier to read.

=cut

use base 'Exporter';
use constant {
  DEBUG => $ENV{ANYEVENT_TEST_MOCK_SERVER_DEBUG}
};
use AnyEvent;
use AnyEvent::Socket;
use AnyEvent::Handle;
use Test::More;
use Scalar::Util qw/weaken/;

our %EXPORT_TAGS = ( 'all' => [ qw(mocksend mockrecv mocksleep mockcode) ] );
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
our @EXPORT = qw();

sub new {
  my $pkg = shift;
  my $self =
    {
     connections => [ @_ ],
     listening => AnyEvent->condvar,
    };
  bless $self, $pkg;
  my $weak_self = $self; weaken $self;
  $self->{server} =
    tcp_server '127.0.0.1', undef, sub {
      my ($fh, $host, $port) = @_;
      print STDERR "In server\n" if DEBUG;
      my $handle;
      $handle =
        AnyEvent::Handle->new(fh => $fh,
                              on_error => sub {
                                my ($hdl, $fatal, $msg) = @_;
                                warn "error $msg\n";
                                $hdl->destroy;
                              },
                              on_eof => sub {
                                my ($hdl) = @_;
                                $hdl->destroy; # destroy handle
                                warn "done.\n";
                              },
                              timeout => 2,
                              on_timeout => sub {
                                die "server timeout\n";
                              }
                             );
      print STDERR "Connection handle: $handle\n" if DEBUG;
      $self->{handles}->{$handle} = $handle;
      my $con = $self->{connections};
      unless (@$con) {
        die "Server received unexpected connection\n";
      }
      my $actions = shift @$con;
      unless (@$con) {
        delete $self->{server};
      }
      $self->next_action($handle, $actions);
    }, sub {
      my ($fh, $host, $port) = @_;
      die "tcp_server setup failed: $!\n" unless ($fh);
      $self->{listening}->send([$host, $port]);
    };
  return $self;
}

sub DESTROY {
  my $self = shift;
  delete $self->{listening};
  delete $self->{server};
  foreach (values %{$self->{handles}}) {
    next unless (defined $_);
    $_->destroy;
    delete $self->{handles}->{$_};
  }
}

sub listening {
  shift->{listening};
}

sub connect_address {
  @{shift->listening->recv};
}

sub connect_host {
  shift->listening->recv->[0];
}

sub connect_port {
  shift->listening->recv->[1];
}

sub connect_string {
  join ':', @{shift->connect_address}
}

sub next_action {
  my ($self, $handle, $actions) = @_;
  print STDERR 'In handle connection ', scalar @$actions, "\n" if DEBUG;
  my $action = shift @$actions;
  unless (defined $action) {
    print STDERR "closing connection\n" if DEBUG;
    $handle->push_shutdown;
    delete $self->{handles}->{$handle};
    return;
  }
  $action->($self, $handle, $actions);
}

sub mocksend {
  my ($data, $desc) = @_;
  sub {
    my ($server, $handle, $actions) = @_;
    my $send = $data;
    $send =~ s/\s+//g if (defined $send);
    print STDERR 'Sending: ', $send, ' ', $desc, "\n" if DEBUG;
    $send = pack 'H*', $send;
    print STDERR 'Sending ', length $send, " bytes\n" if DEBUG;
    $handle->push_write($send);
    $server->next_action($handle, $actions);
  }
}

sub mockrecv {
  my ($data, $desc) = @_;
  sub {
    my ($server, $handle, $actions) = @_;
    my $recv = $data;
    $recv =~ s/\s+//g;
    my $expect = $recv;
    print STDERR 'Waiting for ', $recv, ' ', $desc, "\n" if DEBUG;
    my $len = .5*length $recv;
    print STDERR 'Waiting for ', $len, " bytes\n" if DEBUG;
    $handle->push_read(chunk => $len,
                       sub {
                         my ($hdl, $data) = @_;
                         print STDERR "In receive handler\n" if DEBUG;
                         my $got = uc unpack 'H*', $data;
                         is($got, $expect,
                            '... correct message received by server - '.$desc);
                         $server->next_action($hdl, $actions);
                         1;
                       });
  }
}

sub mocksleep {
  my ($interval, $desc) = @_;
  sub {
    my ($server, $handle, $actions) = @_;
    print STDERR 'Sleeping for ', $interval, ' ', $desc, "\n" if DEBUG;
    my $w;
    $w = AnyEvent->timer(after => $interval,
                         cb => sub {
                           $server->next_action($handle, $actions);
                           undef $w;
                         });
  }
}

sub mockcode {
  my ($code, $desc) = @_;
  sub {
    my ($server, $handle, $actions) = @_;
    $code->($server, $handle);
    $server->next_action($handle, $actions);
  }
}

1;


misc/t/Tester.pm
package # Hide from PAUSE
        Tester;
use strict;
use warnings;
use constant {
  SERVER => $ENV{ANYEVENT_MQTT_SERVER} || 'localhost',
  JOBS => $ENV{ANYEVENT_MQTT_TESTER_JOBS} || 1,
  REPEAT => $ENV{ANYEVENT_MQTT_TESTER_REPEAT} || 1,
  TIMEOUT => $ENV{ANYEVENT_MQTT_TESTER_TIMEOUT} || 5,
  DIAG => $ENV{ANYEVENT_MQTT_TESTER_DIAG},
};

use Test::More;
use AnyEvent::MQTT;
use Net::MQTT::Constants;
use JSON;
use File::Slurp;
use Test::SharedFork;

sub run {
  my ($pkg, $file) = @_;

  my $json = JSON->new;
  my $data = read_file($file);
  $data = $json->decode($data);
  my $conf = $data->{config} || {};
  my $streams = $data->{streams} || [ $data->{stream} ];
  my $logs = $data->{logs} || [ $data->{log} ];
  $conf->{jobs} ||= JOBS;
  my $new = [];
  push @$new, @$streams foreach (1..$conf->{jobs});
  $streams = $new;
  $new = [];
  push @$new, @$logs foreach (1..$conf->{jobs});
  $logs = $new;

  $conf->{topic} ||= '/zqk/test';
  $conf->{host} ||= SERVER;
  $conf->{repeat} ||= REPEAT;
  $conf->{timeout} ||= TIMEOUT * $conf->{repeat} * $conf->{jobs};
  my ($test) = ($0 =~ m!([^/]+)\.t$!);
  $conf->{testname} ||= $test;

  my $timeout = AnyEvent->timer(after => $conf->{timeout},
                                cb => sub { die "timeout\n" });

  foreach my $n (0..($conf->{repeat}-1)) {

    my @pids;
    foreach my $i (0..(@$streams-1)) {
      my $pid = fork;
      die "Fork failed\n" unless (defined $pid);
      if ($pid) {
        push @pids, $pid;
        next;
      }
      #diag('child '.$i);
      my @log;
      $conf->{pid} = $i;
      $conf->{testname} .= '.'.$n.'.'.$i;
      $conf->{topicpid} = $conf->{topic}.'/'.$i;
      run_stream($conf, $streams->[$i], \@log);
      check_log($conf, $logs->[$i], \@log);
      #diag('child '.$i.' finished');
      exit;
    }

    foreach my $pid (@pids) {
      #diag('waiting for child '.$pid);
      waitpid($pid, 0);
      if ($?) {
        die "child died: ", ($?>>8), "\n";
      }
    }
  }
  done_testing();
}

sub run_stream {
  my ($conf, $stream, $log) = @_;
  my $cv;
  my $mqtt;
  my %cv = ();
  my %timer = ();
  my $index = 0;
  foreach my $index (0..-1+@$stream) {
    $conf->{index} = $index;
    my $rec = $stream->[$index];
    my $name =
      $rec->{name} || $index.':'.($rec->{action}||'item').'/'.$conf->{pid};
    my $args = $rec->{arguments} || {};
    $_ = replace_conf($_, $conf) foreach (values %$args);
    if ($rec->{action} eq 'connect') {
      $mqtt = AnyEvent::MQTT->new(host => $conf->{host},
                                  client_id => $conf->{testname},
                                  %$args,
                                  message_log_callback => sub {
                                    push @$log, $_[0].' '.$_[1]->string;
                                  },
                                  on_error => sub {
                                    warn $_[1], "\n";
                                    die "\n" if ($_[0]);
                                  },
                                 );
      ok($cv = $mqtt->connect, 'connect - '.$name);
      ok($cv->recv, '...connected - '.$name)
        or BAIL_OUT('connect failed');
    } elsif ($rec->{action} eq 'subscribe') {
      my $cvname = $rec->{cvname}||$name;
      $cv{$cvname} = AnyEvent->condvar;
      my %args =
        (
         topic => $conf->{topicpid},
         qos => MQTT_QOS_AT_MOST_ONCE,
         callback => sub { $cv{$cvname}->send($_[2]); },
         %$args,
        );

      ok($cv = $mqtt->subscribe(%args), '...subscribe - '.$name);
      is($cv->recv, $rec->{result}, '...subscribed - '.$name);
    } elsif ($rec->{action} eq 'unsubscribe') {
      my %args =
        (
         topic => $conf->{topicpid},
         %$args,
        );

      ok($cv = $mqtt->unsubscribe(%args), '...unsubscribe - '.$name);
      is($cv->recv, $rec->{result}, '...unsubscribed - '.$name);
    } elsif ($rec->{action} eq 'publish') {
      my %args =
        (
         topic => $conf->{topicpid},
         qos => MQTT_QOS_AT_MOST_ONCE,
         message => '',
         %$args,
        );
      ok($cv = $mqtt->publish(%args), '...publish - '.$name);
      ok($cv->recv, '...published - '.$name);
    } elsif ($rec->{action} eq 'wait') {
      my $msg = $cv{$rec->{for}}->recv;
      $cv{$rec->{for}} = AnyEvent->condvar;
      my $result = $rec->{result};
      if (ref $result) {
        foreach my $k (sort keys %$result) {
          is($msg->$k, replace_conf($result->{$k}, $conf),
             '...result '.$k.' - '.$name);
        }
      } else {
        is($msg, replace_conf($result, $conf),
           '...result '.$result.' - '.$name);
      }
    } elsif ($rec->{action} eq 'timeout') {
      my $cvname = $rec->{cvname}||$name;
      $cv{$cvname} = AnyEvent->condvar unless (exists $cv{$cvname});
      $timer{$name} = AnyEvent->timer(after => $rec->{timeout},
                      cb => sub { $cv{$cvname}->send("timeout") });
    } elsif ($rec->{action} eq 'send') {
      ok($cv = $mqtt->_send(%$args, cv => AnyEvent->condvar),
         '...send - '.$name);
      ok($cv->recv, '...sent - '.$name);
      my $cvname = $rec->{cvname}||$name;
      $cv{$cvname} = AnyEvent->condvar;
      my $callback = 'before_'.($rec->{response}||'msg').'_callback';
      $mqtt->{$callback} =
        sub {
          $cv{$cvname}->send($_[0]);
          delete $mqtt->{$callback};
        };
    } else {
      die "Invalid action: ", $rec->{action}, "\n";
    }
  }
}

sub check_log {
  my ($conf, $expected, $log) = @_;
  my $i = 0;
  while (my $m = shift @$expected) {
    my ($str) = ($m =~ m!^(.*?)/!);
    if (ref $m) {
      foreach my $alt (@$m) {
        my $re = $alt->{re};
        if (!defined $re || $log->[0] =~ m!$re!) {
          diag($alt->{diag}) if (DIAG && exists $alt->{diag});
          return check_log($conf, $alt->{log}, $log);
        }
      }
      die "Didn't match any alternative message log pattern\n";
    } else {
      my $got = shift @$log;
      is($got, replace_conf($m, $conf), 'message '.$i.' '.$str);
    }
  } continue {
    $i++;
  }
  is(@$log, 0, 'no extra messages') or
    diag("Got:\n  ", (join "\n  ", @$log), "\n");
}

sub replace_conf {
  my ($m, $conf) = @_;
  foreach my $k (keys %$conf) {
    $m =~ s/\%$k\%/$conf->{$k}/eg;
  }
  $m;
}

1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.012.
use strict;
use warnings;



use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "AnyEvent module for an MQTT client",
  "AUTHOR" => "Mark Hindess <soft-cpan\@temporalanomaly.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "AnyEvent-MQTT",
  "EXE_FILES" => [
    "bin/anyevent-mqtt-monitor",
    "bin/anyevent-mqtt-pub",
    "bin/anyevent-mqtt-sub"
  ],
  "LICENSE" => "perl",
  "NAME" => "AnyEvent::MQTT",
  "PREREQ_PM" => {
    "AnyEvent" => 0,
    "AnyEvent::Handle" => 0,
    "Carp" => 0,
    "Getopt::Long" => 0,
    "Net::MQTT::Constants" => 0,
    "Net::MQTT::Message" => "1.112320",
    "Net::MQTT::TopicStore" => 0,
    "POSIX" => 0,
    "Pod::Usage" => 0,
    "Scalar::Util" => 0,
    "Sub::Name" => 0,
    "constant" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "AnyEvent::MockTCPServer" => 0,
    "AnyEvent::Socket" => 0,
    "Exporter" => 0,
    "File::Temp" => 0,
    "Test::More" => 0,
    "base" => 0
  },
  "VERSION" => "1.212810",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "AnyEvent" => 0,
  "AnyEvent::Handle" => 0,
  "AnyEvent::MockTCPServer" => 0,
  "AnyEvent::Socket" => 0,
  "Carp" => 0,
  "Exporter" => 0,
  "File::Temp" => 0,
  "Getopt::Long" => 0,
  "Net::MQTT::Constants" => 0,
  "Net::MQTT::Message" => "1.112320",
  "Net::MQTT::TopicStore" => 0,
  "POSIX" => 0,
  "Pod::Usage" => 0,
  "Scalar::Util" => 0,
  "Sub::Name" => 0,
  "Test::More" => 0,
  "base" => 0,
  "constant" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
